tools/ci-operator-prowgen/main.go

```diff
@@ -0,0 +1,236 @@
+	flag.BoolVar(&opt.verbose, "v", false, "Show verbose output")
+
+	return opt
+}
+
+func generatePodSpec(org, repo, branch string) *kubeapi.PodSpec {
+	configMapKeyRef := kubeapi.EnvVarSource{
+		ConfigMapKeyRef: &kubeapi.ConfigMapKeySelector{
+			LocalObjectReference: kubeapi.LocalObjectReference{
+				Name: fmt.Sprintf("ci-operator-%s-%s", org, repo),
+			},
+			Key: fmt.Sprintf("%s.json", branch),
+		},
+	}
+
+	return &kubeapi.PodSpec{
+		ServiceAccountName: "ci-operator",
+		Containers: []kubeapi.Container{
+			kubeapi.Container{
+				Name:    "test",
+				Image:   "ci-operator:latest",
+				Command: []string{"ci-operator"},
+				Args:    []string{"--artifact-dir=$(ARTIFACTS)"},
+				Env: []kubeapi.EnvVar{
+					kubeapi.EnvVar{
+						Name:      "CONFIG_SPEC",
+						ValueFrom: &configMapKeyRef,
+					},
+				},
+			},
+		},
+	}
+}
+
+type testDescription struct {
+	Name   string
+	Target string
+}
+
+func generatePresubmitForTest(test testDescription, org, repo, branch string) *prowconfig.Presubmit {
+	presubmit := prowconfig.Presubmit{
+		Agent:        "kubernetes",
+		AlwaysRun:    true,
+		Brancher:     prowconfig.Brancher{Branches: []string{branch}},
+		Context:      fmt.Sprintf("ci/prow/%s", test.Name),
+		Name:         fmt.Sprintf("pull-ci-%s-%s-%s", org, repo, test.Name),
+		RerunCommand: fmt.Sprintf("/test %s", test.Name),
+		// Spec:         generatePodSpec(org, repo, branch),
+		Trigger: fmt.Sprintf("((?m)^/test( all| %s),?(\\\\s+|$))", test.Name),
+		UtilityConfig: prowconfig.UtilityConfig{
+			DecorationConfig: &prowkube.DecorationConfig{SkipCloning: true},
+			Decorate:         true,
+		},
+	}
+	presubmit.Spec.Containers[0].Args = append(
+		presubmit.Spec.Containers[0].Args,
+		fmt.Sprintf("--target=%s", test.Target),
+	)
+
+	return &presubmit
+}
+
+func generatePostsubmitForTest(test testDescription, org, repo, branch string) *prowconfig.Postsubmit {
+	postsubmit := prowconfig.Postsubmit{
+		Agent: "kubernetes",
+		Name:  fmt.Sprintf("branch-ci-%s-%s-%s", org, repo, test.Name),
+		// Spec:  generatePodSpec(org, repo, branch),
+		UtilityConfig: prowconfig.UtilityConfig{
+			DecorationConfig: &prowkube.DecorationConfig{SkipCloning: true},
+			Decorate:         true,
+		},
+	}
+
+	postsubmit.Spec.Containers[0].Args = append(
+		postsubmit.Spec.Containers[0].Args,
+		fmt.Sprintf("--target=%s", test.Target),
+	)
+	postsubmit.Spec.Containers[0].Args = append(postsubmit.Spec.Containers[0].Args, "--promote")
+
+	return &postsubmit
+}
+
+func generateJobs(
+	configSpec *cioperatorapi.ReleaseBuildConfiguration,
+	org, repo, branch string,
+) (*map[string][]prowconfig.Presubmit, *map[string][]prowconfig.Postsubmit) {
+
+	orgrepo := fmt.Sprintf("%s/%s", org, repo)
+	presubmits := map[string][]prowconfig.Presubmit{}
+	postsubmits := map[string][]prowconfig.Postsubmit{}
+
+	imagesTest := false
+
+	for _, element := range configSpec.Tests {
+		// Check if config file has "images" test defined to avoid name clash
```

Comment:
NVM figured it out

Reply:
I'm taking of the special case where someone would have `images` test in their config file. The generator adds a job with `--target=[images]` parameter in addition to the jobs derived from `tests` keys. The generator derives names from the test identifier, but the job name cannot be `...-[images]` so the generator calls such job `...--images` which would clash with a job made for `images` test in the config file.

But... while writing this, I realized - is it really the case? Maybe I was just too careful and it is not actually needed. @stevekuznetsov, do you think we could simply use `[images]` there? It would mean we have job called `...-[images]` and stuff like `context: ci/prow/[images]` and `rerun_command: /test [images]`. It would make the generator code simpler.

Reply:
I think even if it is allowed, hiding the `[` on `images` should be something we do, it's cleaner for the end user. The `[images]` and promotion tasks are more admin tasks than tests so we can own the generation of these 100% in this tool.